src/main/java/tfb/status/service/ShareResultsUploader.java

```diff
@@ -0,0 +1,305 @@
+import java.util.regex.Pattern;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.io.ByteStreams;
+import com.google.common.io.MoreFiles;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import tfb.status.config.FileStoreConfig;
+import tfb.status.config.UrlsConfig;
+import tfb.status.util.ZipFiles;
+import tfb.status.view.Results;
+import tfb.status.view.ShareResultsJsonView;
+
+import static com.google.common.net.MediaType.JSON_UTF_8;
+import static io.undertow.util.Headers.CONTENT_TYPE;
+import static io.undertow.util.StatusCodes.NOT_FOUND;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.nio.file.StandardOpenOption.APPEND;
+import static java.nio.file.StandardOpenOption.READ;
+import static java.nio.file.StandardOpenOption.WRITE;
+import static java.util.regex.Pattern.CASE_INSENSITIVE;
+
+/**
+ * A service responsible for handling public uploads of results.json files.
+ * This class provides several ways of creating new shared results files in
+ * the {@link FileStore#shareDirectory()}. All access to that directory
+ * should be done through this service.
+ */
+@Singleton
+public class ShareResultsUploader {
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+  private final FileStoreConfig fileStoreConfig;
+  private final UrlsConfig urlsConfig;
+  private final FileStore fileStore;
+  private final ObjectMapper objectMapper;
+
+  @Inject
+  public ShareResultsUploader(FileStoreConfig fileStoreConfig,
+                              UrlsConfig urlsConfig,
+                              FileStore fileStore,
+                              ObjectMapper objectMapper) {
+    this.fileStoreConfig = Objects.requireNonNull(fileStoreConfig);
+    this.urlsConfig = Objects.requireNonNull(urlsConfig);
+    this.fileStore = Objects.requireNonNull(fileStore);
+    this.objectMapper = Objects.requireNonNull(objectMapper);
+  }
+
+  /**
+   * Upload the given raw results JSON string. Creates a {@link ReadableByteChannel}
+   * from the string and passes it to {@link #upload(ReadableByteChannel)}.
+   *
+   * @param resultsJson The raw results JSON.
+   * @see #upload(ReadableByteChannel)
+   */
+  public ShareResultsJsonView upload(String resultsJson)
+          throws IOException, ShareResultsUploadException {
+    Objects.requireNonNull(resultsJson);
+
+    try (InputStream is = new ByteArrayInputStream(resultsJson.getBytes(UTF_8));
+         ReadableByteChannel in = Channels.newChannel(is)) {
+      return upload(in);
+    }
+  }
+
+  /**
+   * Upload the given byte channel containing the raw results JSON. Creates
+   * a temporary file from the bytes and passes it to {@link #upload(Path)}.
```

Comment:
This project uses an 80 character margin.  Here, there is room for "a" to fit on the previous line.

I'm not particularly concerned with this "a", but I notice a lot of cases like this.  Are you eyeballing the margin?  If so, I suggest using IntelliJ's "visual guides" feature.  Editor - Code Style - General - Visual guides.  I have mine set to "80, 100, 120" since those are common margins used in projects.  If you find the guides unpleasant and distracting, I still suggest enabling them while working on this project at least to help keep the wrapping correct.